src/main/java/org/commcare/resources/model/ResourceTable.java

```diff
@@ -1169,14 +1163,19 @@ public boolean recoverResources(CommCarePlatform platform, String profileRef, Ve
     }
 
     // Downloads and re-installs a missing resource
-    public void recoverResource(Resource missingResource, CommCarePlatform platform, String profileRef)
+    public void recoverResource(Resource missingResource, CommCarePlatform platform, String profileRef, InstallRequestSource source)
             throws InstallCancelledException, UnresolvedResourceException, UnfullfilledRequirementsException {
 
         if (missingResource.id.contentEquals(CommCarePlatform.APP_PROFILE_RESOURCE_ID)) {
             addRemoteLocationIfMissing(missingResource, profileRef);
         }
 
-        findResourceLocationAndInstall(missingResource, new Vector<>(), false, platform, null, true);
+        try {
+            platform.registerInstallContext(new ResourceInstallContext(source));
```

Comment:
I have mixed feelings about this approach. I think I'd default to passing through at the method level rather than making the Platform object stateful in this way, since it makes the platform object much, much more complex (and it makes it hard for someone to identify that the install context is available / required and hidden inside the platform unless you happen to know) but non-blocking as feedback.

Reply:
Agree that this design does feel bad. As an alternative do you mean have `ResourceInstallContext` as a method parameter in `ResourceInstaller::install` right ?

Reply:
yeah, I'm not sure if there are limitations there which I'm missing, but I'm imagining it replacing the existing `upgrade` boolean flag, providing a single place where new context fields (like that upgrade boolean) can be expanded without us needing to change all of these signatures again in the future.

Reply:
@ctsims  merging this after implementing this suggestion, though feel free to have a final review anyway and leave feedback if any.